syntax = "proto3";

package raft;

service Raft {
  // AppendEntries is used to replicate log entries to followers.
  rpc AppendEntries(AppendEntriesRPC) returns (AppendEntriesReply);

  // RequestVote is used by candidates to gather votes from other nodes.
  rpc RequestVote(RequestVoteRPC) returns (RequestVoteReply);
}

message RequestVoteRPC {
    int32 term = 1;
    string candidate_id = 2;
    int32 last_log_index = 3;
    int32 last_log_term = 4;
}

message RequestVoteReply {
    int32 term = 1;
    bool vote_granted = 2;
}

message LogEntry {
    int32 term = 1;
    repeated string command = 2; // Command to be executed
}

message AppendEntriesRPC {
    int32 term = 1; 
    string leader_id = 2; 
    int32 prev_log_index = 3; 
    int32 prev_log_term = 4; 
    repeated LogEntry entries = 5; 
    int32 leader_commit = 6;
}

message AppendEntriesReply {
    int32 term = 1; 
    bool success = 2; 
}